tools/link.py

```diff
@@ -2335,7 +2298,38 @@ def node_es6_imports():
 '''
 
 
-def modularize():
+def make_pthread_detection():
+  """Create code for detecting if we are running in a pthread.
+
+  Normally this detection is done when the module is itself is run but there
+  are some cases were we need to do this detection outside of the normal
+  module code.
+
+  1. When running in MODULARIZE mode we need use this to know if we should
+     run the module constructor on startup (true only for pthreads)
+  2. When using `--extern-pre-js` and `--extern-post-js` we need to avoid
+     running this code on pthreads.
+  """
+
+  if settings.ENVIRONMENT_MAY_BE_WEB or settings.ENVIRONMENT_MAY_BE_WORKER:
+    code = "var isPthread = globalThis.self?.name === 'em-pthread';\n"
+    if settings.ENVIRONMENT_MAY_BE_NODE:
+      code += "var isNode = typeof globalThis.process?.versions?.node == 'string';\n"
+      code += "if (isNode) isPthread ="
+  elif settings.ENVIRONMENT_MAY_BE_NODE:
+    code = "var isPthread ="
+  else:
+    assert False
+
+  if settings.ENVIRONMENT_MAY_BE_NODE:
+    if settings.EXPORT_ES6:
+      code += "(await import('worker_threads')).workerData;\n"
+    else:
+      code += "require('worker_threads').workerData\n"
```

Comment:
This just checks the environment may be node, but it may be a build that also runs on the web? In that case how would `require` work?

Reply:
Tn that case this line would be preceded by `if (isNode) isPthread =`.   I agree its hard to follow.  We can try to find a more readable way to keep this concise perhaps?

Reply:
I see, thanks. Code looks ok, a little hard to read but I can't think of an obvious improvement, I was just reading it wrong before.

Reply:
I updated to make to more clear what is going on here.